model/riscv_insts_zawrs.sail

```diff
@@ -0,0 +1,118 @@
+/*     notice, this list of conditions and the following disclaimer.                     */
+/*  2. Redistributions in binary form must reproduce the above copyright                 */
+/*     notice, this list of conditions and the following disclaimer in                   */
+/*     the documentation and/or other materials provided with the                        */
+/*     distribution.                                                                     */
+/*                                                                                       */
+/*  THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''                   */
+/*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED                    */
+/*  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A                      */
+/*  PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR                  */
+/*  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,                         */
+/*  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT                     */
+/*  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF                     */
+/*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND                  */
+/*  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,                   */
+/*  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT                   */
+/*  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF                   */
+/*  SUCH DAMAGE.                                                                         */
+/*=======================================================================================*/
+
+/* ****************************************************************** */
+/* This file specifies the instructions in the 'Zawrs' extension.  */
+
+val reservation_set_valid = {
+  ocaml:       "Platform.reservation_set_valid",
+  interpreter: "Platform.reservation_set_valid",
+  c:           "reservation_set_valid",
+  lem:         "reservation_set_valid"
+  } : unit -> bool
+
+/* ****************************************************************** */
+union clause ast = WRS : (wrsop)
+
+mapping clause encdec = WRS(WRS_STO) if haveZawrs()
+  <-> 0b000000011101 @ 0b00000 @ 0b000 @ 0b00000 @ 0b1110011 if haveZawrs()
+
+mapping clause encdec = WRS(WRS_NTO) if haveZawrs()
+  <-> 0b000000001101 @ 0b00000 @ 0b000 @ 0b00000 @ 0b1110011 if haveZawrs()
+
+/* Hart must resume if a locally enabled interrupt becomes pending, even if delegated */
+function interrupts_pending() -> bool = ((mip.bits() & mie.bits()) != zero_extend(0b0))
+
+function clause execute (WRS(op)) = {
+
+  platform_start_wrs_timeout(if op == WRS_STO then true else false);
+
+  while reservation_set_valid() & not(interrupts_pending()) & not(platform_is_wrs_timed_out()) do ();
```

Comment:
I'm not sure if the best way to handle this is via waiting in the actual execute function. WFI doesn't work like this (to be fair it doesn't work at all because there's no proper interface for injecting interrupts currently).

Maybe the model should be a state machine and the `step()` function handled the WFI and WRS states explicitly?

I dunno, we probably need to think about this properly. Maybe this is fine for now though since the platform callbacks are just hard-coded to time out immediately.

Reply:
The intent here is two folds:
1. As a precise documentation of the ISA - this shows that the execution stalls while the instruction waits for these three conditions a) a snoop that invalidates the reservation set due to a store from another hart or a device b) an interrupt becomes pending c) the wait times out.
2. Allows the platform test bench to inject these into the mode - a snoop invalidation or the platform specific timer timeout. This would also include advancing cycles and time.

We cannot `step()` as it is presently as we cannot fetch or execute instructions when stalled.

Reply:
Yeah I agree it is nice to document the stalls like this. But documentation isn't the only purpose of the model. I think we want to be a bit careful about this because this is the first bit of code in the model that assumes that `step()` can block.

> We cannot step() as it is presently as we cannot fetch or execute instructions when stalled.

My feeling is that `step()` should really mean "step the model state", not "fetch and execute an instruction". At the moment those are exactly the same because none of the execute functions call platform callbacks to wait for an event, including WFI. If `step()` blocks you need threads in lots of situations (e.g. multiple harts, cosimulation), which are quite a pain in C++ and even worse in C.

But it's true that turning `step()` into an explicit state machine would make the code less clear. Hmm.

Reply:
Thats why I wrote "as it is presently". I dont think it would make the code less clear as a state machine. We would need instead of RETIRE_SUCCESS, something like STALL_UNTIL and the code should setup the conditions for "until". The model would then want the execution to resume after the STALL_UNTIL - STALL_UNTIL then becomes sort of like a `swapcontext()`.

Reply:
> there's no proper interface for injecting interrupts currently

#395 provides timer interrupts.